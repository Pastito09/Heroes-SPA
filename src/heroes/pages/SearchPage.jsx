import { useNavigate, useLocation } from "react-router-dom";
import queryString from 'query-string';


import { useForm } from "../../hooks/useForm";
import { HeroCard } from "../components/index";
import { getHeroesByName } from "../helpers";
//import { require } from "node:module";

//const querystring = require("querystring");

export const SearchPage = () => {
  
  const navigate = useNavigate();
  const location = useLocation();

  const { q = "" } = queryString.parse( location.search );
  const heroes = getHeroesByName( q );

  const showSearch = (q.length === 0);//ya son valores booleanos
  const showError = (q.length > 0) && heroes.length === 0; //booleano

  const { searchText, onInputChange } = useForm({
    searchText: q,
  });

  const onSearchSubmit = ( event ) => { 
    event.preventDefault();
    //if ( searchText.trim().length <= 1 ) return;

    navigate(`?q=${ searchText }`);

  };

  return (
    <>
    <h1>PÃ¡gina de busqueda</h1>
    <hr />

    <div className="row">
      
      <div className="col-5">
        <h4>Buscando</h4>
          <hr />
          <form onSubmit={ onSearchSubmit } aria-label="form">
            <input 
              type="text"
              placeholder="Busca un heroe"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={ searchText }
              onChange={ onInputChange }           
            />

            <button className="btn btn-outline-primary mt-1">
              Busqueda
            </button>
          </form>
      </div>

      <div className="col-7">
          <h4>Resultados</h4>
          <hr />
        {/* {
          ( q === '' )
           ?  <div className="alert alert-primary">Busca un heroe</div>
          : ( heroes.length === 0 ) 
            && <div className="alert alert-danger">No hay resultados con <b>{ q }</b></div>
         } */}
         
        <div /* otra forma de hacer la condicional de mostrar segun el resultado*/
        className="alert alert-primary animate__animated animate__fadeIn" 
        style={{display: showSearch ? '' : 'none' }}
          >
          Busca un heroe
        </div>
        
        <div 
        aria-label="alert-danger"
        className="alert alert-danger animate__animated animate__fadeIn"
        style={{display: showError ? '' : 'none' }}
        >
          No hay resultados con <b>{ q }</b>
        </div>
        
        
        {
          heroes.map( hero => (
            <HeroCard key={ hero.id } { ...hero } />
          ))
        }

    </div>

</div>
    
    </>
  )
}
