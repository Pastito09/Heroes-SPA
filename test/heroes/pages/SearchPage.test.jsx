import { fireEvent, render, screen } from "@testing-library/react";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";
import { MemoryRouter } from "react-router-dom";

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    mockedUseNavigate: () => mockedUseNavigate,
}));

describe('pruebas en <SearchPage />', () => { 

    beforeEach(() => jest.clearAllMocks());
    
        test('debe de mostrarse correctamente con valores por defecto', () => { 
        

            const { container } = render(
                <MemoryRouter>
                    <SearchPage />

                </MemoryRouter>
     
                );
            
            expect( container ).toMatchSnapshot();
            //screen.debug()
     });

     test('debe de mostrar a Batman y el input con el valor del queryString', () => { 
        

        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
            );
        
        const input = screen.getByRole("textbox")
        //expect( input.value ).toBe('batman');

        //const img = screen.getByRole("img");
        //expect( img.src ).toContain('/public/heroes/dc-batman.jpg');

        const alert = screen.getByLabelText("alert-danger");
        //expect( alert.style.display ).toBe('none');

 });

    test('debe de mostrar un erro si no se encuentra el heroe (batman123)', () => {

        render(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchPage />
            </MemoryRouter>
            );

        const alert = screen.getByLabelText("alert-danger");
        //expect( alert.style.display ).toBe('')

    });

    test('debe de llamar el navigate a la pantalla nueva', () => { 
        
        const inputValue = 'superman';

        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage />
            </MemoryRouter>
            );
            
            const input = screen.getByRole('textbox');
            fireEvent.change(input, { target: {name: 'searchText', value: inputValue }});
            //console.log(input.value)

            const form = screen.getByRole('form');
            fireEvent.submit(form);

            expect( mockedUseNavigate ).toHaveBeenCalledWith(`?=${ inputValue }`);
     });
     
 });