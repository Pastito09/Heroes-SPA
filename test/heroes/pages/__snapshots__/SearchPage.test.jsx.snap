// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <SearchPage /> debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    PÃ¡gina de busqueda
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Buscando
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Busca un heroe"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Busqueda
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Busca un heroe
      </div>
      <div
        aria-label="alert-danger"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No hay resultados con 
        <b />
      </div>
    </div>
  </div>
</div>
`;
